sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/json/JsonSuite.scala

```diff
@@ -963,6 +963,31 @@ class JsonSuite extends QueryTest with SharedSQLContext with TestJsonData {
     )
   }
 
+  test("SPARK-13764 Parse modes in JSON data source") {
+    withSQLConf(SQLConf.COLUMN_NAME_OF_CORRUPT_RECORD.key -> "_unparsed") {
+      // `FAILFAST` mode should throw an exception for corrupt records.
+      val exception = intercept[SparkException] {
+        sqlContext.read.option("mode", "FAILFAST").json(corruptRecords).collect()
+      }
+      assert(exception.getMessage.contains("Malformed line in FAILFAST mode: {"))
+
+      // `DROPMALFORMED` mode should skip corrupt records
+      // For `PERMISSIVE` mode, it is tested in "Corrupt records" test.
+      val jsonDF = sqlContext.read.option("mode", "DROPMALFORMED").json(corruptRecords)
+      val schema = StructType(
+        StructField("_unparsed", StringType, true) ::
+          StructField("a", StringType, true) ::
+          StructField("b", StringType, true) ::
+          StructField("c", StringType, true) :: Nil)
+      assert(schema === jsonDF.schema)
+
+      checkAnswer(
+        jsonDF,
+        Row(null, "str_a_4", "str_b_4", "str_c_4") :: Nil
```

Comment:
I realised that It will has a `_corrupt_record` field as it fails to parse corrupt records. So, I thought it would be better just  to show `_unparsed` explicitly rather than using the defualt `_corrupt_record`.